### PR TITLE
[TorchToLinalg] Lower aten.__and__.Tensor with tensor broadcasting

### PR DESCRIPTION
**Pull request by weimin023**
## Motivation

This was found while compiling a text-only Turbine export of [gemma-4-31B](https://huggingface.co/google/gemma-4-31B) with IREE.

The exported boolean attention-mask path contains:

```
%true = torch.vtensor.literal(dense<true> : tensor<i1>)
    : !torch.vtensor<[],i1>
%mask = ... : !torch.vtensor<[1,1,3,3],i1>
%result = torch.aten.__and__.Tensor %true, %mask
    : !torch.vtensor<[],i1>, !torch.vtensor<[1,1,3,3],i1>
    -> !torch.vtensor<[1,1,3,3],i1>
```

## Summary

Add TorchToLinalg support for `torch.aten.__and__.Tensor`.

PyTorch exports Python tensor `&` expressions as `aten.__and__.Tensor` in some graphs. 
TorchToLinalg already lowers `aten.bitwise_and.Tensor`, but did not recognize this distinct Torch dialect operation.
As a result, `aten.__and__.Tensor` remained in the IR and blocked final type conversion.

This change reuses the existing bitwise-AND elementwise lowering:

- Recognize `Aten__And__TensorOp` in `ConvertElementwiseOp`.
- Reuse the `arith::AndIOp` payload used for `AtenBitwiseAndTensorOp`.
- Mark `Aten__And__TensorOp` illegal in the TorchToLinalg conversion target.
- Add a regression for broadcasting a rank-0 `i1` tensor across a ranked `i1` tensor.
- Fix mixed-dtype promotion in `aten.bitwise_and/or/xor.Tensor` (and `aten.__and__.Tensor`) lowering: 
  propagate the original operand dtype through `convertScalarToDtype` so unsigned operands are zero-extended (not sign-extended) when promoted to a wider integer type.

## Reproducer

```mlir
func.func @and_tensor_i1_scalar_broadcast(
    %arg0: !torch.vtensor<[],i1>,
    %arg1: !torch.vtensor<[1,1,3,3],i1>)
    -> !torch.vtensor<[1,1,3,3],i1> {
  %0 = torch.aten.__and__.Tensor %arg0, %arg1
      : !torch.vtensor<[],i1>, !torch.vtensor<[1,1,3,3],i1>
      -> !torch.vtensor<[1,1,3,3],i1>
  return %0 : !torch.vtensor<[1,1,3,3],i1>
}
```

### Lit Test
```
torch-mlir-opt --convert-torch-to-linalg test/Conversion/TorchToLinalg/and_tensor.mlir

```

### Verification (e2e test)
Propagated the original operand dtype through `convertScalarToDtype` so unsigned operands are zero-extended (not sign-extended) on promotion.

Lowered payload now uses arith.extui (zero-extend) instead of arith.extsi, which previously produced [-128, -1] instead of [128, 255].

```
cd projects/pt1
python3 -m e2e_testing.main --config=linalg --filter='ElementwiseAndTensorMixedDtypeModule_basic' -v
```

```
LINALG Backend IR
#map = affine_map<(d0) -> (d0)>
module attributes {torch.debug_module_name = "ElementwiseAndTensorMixedDtypeModule"} {
  func.func @forward(%arg0: tensor<?xi8>, %arg1: tensor<?xi64>) -> tensor<?xi64> {
    %c0 = arith.constant 0 : index
    %dim = tensor.dim %arg0, %c0 : tensor<?xi8>
    %dim_0 = tensor.dim %arg1, %c0 : tensor<?xi64>
    %0 = arith.cmpi eq, %dim, %dim_0 : index
    cf.assert %0, "mismatched size for broadcast"
    %1 = tensor.empty(%dim) : tensor<?xi64>
    %2 = linalg.generic {indexing_maps = [#map, #map, #map], iterator_types = ["parallel"]} ins(%arg0, %arg1 : tensor<?xi8>, tensor<?xi64>) outs(%1 : tensor<?xi64>) {
    ^bb0(%in: i8, %in_1: i64, %out: i64):
      %3 = arith.extui %in : i8 to i64                 <---------- This one
      %4 = arith.andi %3, %in_1 : i64
      linalg.yield %4 : i64
    } -> tensor<?xi64>
    return %2 : tensor<?xi64>
  }
}

Running ElementwiseAndTensorMixedDtypeModule_basic...
TORCH_VERSION_FOR_COMPARISON = 2.11.0
PASS - "ElementwiseAndTensorMixedDtypeModule_basic"

Summary:
    Passed: 1
```

### CODE REVIEW

**Comment by weimin023** [comment]
Hi @sahas3 @Lallapallooza @rengolin, Could you please take a look at this small TorchToLinalg lowering for `aten.__and__.Tensor`? It reuses the existing `aten.bitwise_and.Tensor` elementwise lowering and adds a scalar-boolean broadcast regression. Thx!

**Review by rengolin — COMMENTED** [review]
From a lowering point of view, this looks ok, and the code change is really minimal (most of changes are formatting), so I'm ok with this.

But I'm new to torch-mlir, and would like someone else to check the meaning of the `__and__` Aten operator when compared to `BitwiseAnd`.

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 1]
```diff
@@ -533,15 +533,13 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
...
+    if (isa<mlir::FloatType>(cast<ValueTensorType>(resultType).getDtype())) {
+      op->emitError("Bitwise_And does not support floating point dtype");
       return nullptr;
     }
-    Type dtype = cast<RankedTensorType>(
-                     converter->convertType(bitwiseAndTensor.getType()))
+    Type dtype = cast<RankedTensorType>(converter->convertType(resultType))
                      .getElementType();
     Value lhs = convertScalarToDtype(b, loc, payloadArgs[0], dtype);
     Value rhs = convertScalarToDtype(b, loc, payloadArgs[1], dtype);
```
Comment (Lallapallooza) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp]: This issue is pre-existing in the shared bitwise lowering, but enabling `__and__.Tensor` exposes it here too. Could we pass the original operand/result dtypes through these conversions and add the mixed-dtype regression in this change?

Repro:

```python
import torch
from torch_mlir import fx
from torch_mlir_e2e_test.linalg_on_tensors_backends.refbackend import (
    RefBackendLinalgOnTensorsBackend,
)

class And(torch.nn.Module):
    def forward(self, x, y):
        return x & y

x = torch.tensor([128, 255], dtype=torch.uint8)
y = torch.tensor([-1, -1], dtype=torch.int64)

module = fx.export_and_import(
    And(),
    x,
    y,
    output_type="linalg-on-tensors",
    decomposition_table={},
)
backend = RefBackendLinalgOnTensorsBackend()
compiled = backend.load(backend.compile(module))

expected = (x & y).tolist()
actual = compiled.main(x.numpy(), y.numpy()).tolist()
assert actual == expected, f"{actual=} != {expected=}"
```

**Review by Lallapallooza — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp (hunk 1 above)
Comment (weimin023) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp]: Hi @Lallapallooza, Thanks for the suggestion. The issue is shared by the identical tensor bitwise promotion paths for AND/OR/XOR, so I applied the same dtype propagation fix to all three, with a mixed-dtype regression for each.

**Comment by weimin023** [comment]
@Lallapallooza, I also notice that, the new test is xfailed for fx_importer_tosa — TorchToTosa has a separate promotion path that still sign-extends unsigned operands. Passes on onnx, fx_importer, and fx_importer_stablehlo.
Let me know if you'd prefer the TOSA path fixed here, or I can file a follow-up issue.

**Comment by Lallapallooza** [comment]
> @Lallapallooza, I also notice that, the new test is xfailed for fx_importer_tosa — TorchToTosa has a separate promotion path that still sign-extends unsigned operands. Passes on onnx, fx_importer, and fx_importer_stablehlo. Let me know if you'd prefer the TOSA path fixed here, or I can file a follow-up issue.

it's ok to xfail

**Review by Lallapallooza — APPROVED** [review]
LGTM
